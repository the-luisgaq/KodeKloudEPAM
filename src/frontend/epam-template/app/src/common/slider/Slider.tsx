import * as React from 'react';
import css from './Slider.module.scss';
import { Slide, SlideProps } from './Slide';
import { IconButton } from '@epam/uui';
import { IAnalyticableOnChange, UuiContext, UuiContexts } from '@epam/uui-core';
import { ReactComponent as ArrowPrev } from '../../icons/navigation-left.svg';
import { ReactComponent as ArrowNext } from '../../icons/navigation-right.svg';

export interface SliderProps extends IAnalyticableOnChange<number> {
    slides: SlideProps[] | null;
}

export class Slider extends React.Component<SliderProps> {
    public static contextType = UuiContext;
    public context: UuiContexts;
    constructor(props: SliderProps) {
        super(props);
    }

    state = { activeSlide: 0 };
    handlePreviousClick = () => {
        this.setState({ activeSlide: this.state.activeSlide - 1 });

        if (this.props.getValueChangeAnalyticsEvent) {
            const event = this.props.getValueChangeAnalyticsEvent(this.state.activeSlide - 1, this.state.activeSlide);
            this.context.uuiAnalytics.sendEvent(event);
        }
    };

    handleNextClick = () => {
        this.setState({ activeSlide: this.state.activeSlide + 1 });

        if (this.props.getValueChangeAnalyticsEvent) {
            const event = this.props.getValueChangeAnalyticsEvent(this.state.activeSlide + 1, this.state.activeSlide);
            this.context.uuiAnalytics.sendEvent(event);
        }
    };

    render() {
        const { slides } = this.props;
        return (
            <div className={ css.slider }>
                <div className={ css.controls }>
                    <IconButton
                        rawProps={ { 'aria-label': 'Backward' } }
                        color="primary"
                        isDisabled={ this.state.activeSlide === 0 }
                        icon={ ArrowPrev }
                        onClick={ this.handlePreviousClick }
                    />
                    <IconButton
                        rawProps={ { 'aria-label': 'Forward' } }
                        color="primary"
                        isDisabled={ this.state.activeSlide === this.props.slides.length - 1 }
                        icon={ ArrowNext }
                        onClick={ this.handleNextClick }
                    />
                </div>
                <Slide { ...slides[this.state.activeSlide] } />
            </div>
        );
    }
}
