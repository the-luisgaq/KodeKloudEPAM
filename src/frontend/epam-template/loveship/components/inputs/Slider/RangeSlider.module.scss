@use '../../../assets/styles/scss/index'  as *;

.root {
    :global(.uui-tooltip-body) {
        background-color: $night700;
        padding: 0 10px;
        color: white;
        line-height: 24px;
        font-size: 14px;
        font-family: $font-sans;
        box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.2);
    }

    :global(.uui-tooltip-arrow) {
        border: 5px solid transparent;
    }

    :global(.uui-slider-scale-number) {
        font-size: 11px;
        font-family: $font-sans;
        color: $night700;
        margin-top: 10px;
    }

    :global(.uui-slider-scale-dot) {
        width: 7px;
        height: 7px;
        border-radius: 50%;
        top: -1px;
    }

    &:global(.uui-slider-container) {
        height: 12px;

        :global(.uui-slider) {
            margin: 0 6px 0 6px;
            width: 100%;
            height: 3px;
            border-radius: 10px;
            background: $night500;
            min-width: 100px;
        }

        :global(.uui-slider-scale) {
            height: 6px;
        }

        :global(.uui-slider-handle) {
            width: 12px;
            height: 12px;
            border-radius: 50%;
            background-color: $sky;
            cursor: pointer;
        }

        :global(.uui-slider-filled) {
            margin: 0 6px;
            height: 3px;
            border-radius: 10px;
            background-color: $sky;
        }

        :global(.uui-slider-scale-dot) {
            background-color: $night200;
            border: 1px $night500 solid;
        }

        &:hover {
            :global(.uui-slider) {
                background-color: $sky-light;
            }

            :global(.uui-slider-scale-dot):not(:global(.uui-slider-scale-filled-dot)) {
                border-color: $sky-light;
                background-color: $sky-light;
            }
        }

        :global(.uui-slider-scale-filled-dot) {
            background-color: $sky;
            border-color: $sky;
        }

        &:global(.uui-active) {
            :global(.uui-slider) {
                background-color: $sky-light;
            }

            :global(.uui-slider-scale-dot):not(:global(.uui-slider-scale-filled-dot)) {
                border-color: $sky-light;
                background-color: $sky-light;
            }
        }

        &:global(.uui-disabled) {
            pointer-events: none;

            :global(.uui-slider-filled) {
                background-color: $night400;
            }

            :global(.uui-slider-handle) {
                background-color: $night400;
            }

            :global(.uui-slider) {
                background-color: $night400;
            }

            :global(.uui-slider-scale-filled-dot) {
                background-color: $night400;
                border-color: $night400;
            }
        }
    }
}
