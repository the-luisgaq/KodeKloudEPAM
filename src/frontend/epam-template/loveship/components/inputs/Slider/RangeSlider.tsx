import {
    RangeSlider as uuiRangeSlider, RangeSliderProps as uuiRangeSliderProps,
} from '@epam/uui-components';
import css from './RangeSlider.module.scss';
import { createSkinComponent } from '@epam/uui-core';

/** Represents the properties of a RangeSlider component. */
export interface RangeSliderProps extends uuiRangeSliderProps {}

export function applyRangeSliderMods() {
    return [
        css.root,
    ];
}

export const RangeSlider = createSkinComponent<uuiRangeSliderProps, RangeSliderProps>(
    uuiRangeSlider,
    () => null,
    applyRangeSliderMods,
);
