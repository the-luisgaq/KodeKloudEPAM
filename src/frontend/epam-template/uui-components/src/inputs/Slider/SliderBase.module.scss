.root {
    &:global(.uui-slider-container) {
        display: flex;
        position: relative;
        align-items: center;
        width: 100%;
        padding: 12px 0;
        cursor: pointer;

        :global(.uui-slider-handle) {
            position: absolute;
            left: 0;

            &:focus {
                outline-offset: 1px;
                outline-color: var(--uui-color-700);
            }
        }

        :global(.uui-slider) {
            pointer-events: none;
        }

        :global(.uui-slider-filled) {
            position: absolute;
            pointer-events: none;
        }

        :global(.uui-slider-scale) {
            position: absolute;
        }

        :global(.uui-slider-scale-number) {
            position: absolute;
        }

        :global(.uui-slider-scale-dot) {
            position: absolute;
        }
    }
}


