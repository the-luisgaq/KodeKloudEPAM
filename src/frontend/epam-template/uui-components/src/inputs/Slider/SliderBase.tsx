import * as React from 'react';
import {
    IHasCX, IEditable, IDisableable, IHasRawProps, IHasForwardedRef,
} from '@epam/uui-core';

export interface SliderBaseProps<TSelection>
    extends IHasCX,
    IEditable<TSelection>,
    IDisableable,
    IHasRawProps<React.HTMLAttributes<HTMLDivElement>>,
    IHasForwardedRef<HTMLDivElement> {
    /** Min value (when slider is at leftmost position) */
    min: number;
    /** Max value (when slider is at rightmost position) */
    max: number;
    /** Value change step */
    step: number;

    /** Slider renders a label at each splitAt position */
    splitAt?: number;

    /** Overrides rendering of values labels */
    renderLabel?: (value: number) => string;

    /** Toggles rendering of tooltip with current value */
    showTooltip?: boolean;
}

export interface SliderBaseState {
    isActive: boolean;
    valueWidth: number;
}

export const uuiSlider = {
    container: 'uui-slider-container',
    slider: 'uui-slider',
    filled: 'uui-slider-filled',
    handle: 'uui-slider-handle',
    scale: 'uui-slider-scale',
    scaleNumber: 'uui-slider-scale-number',
    scaleDot: 'uui-slider-scale-dot',
    scaleFilledDot: 'uui-slider-scale-filled-dot',
} as const;

export abstract class SliderBase<TSelection, TState extends SliderBaseState> extends React.Component<SliderBaseProps<TSelection>, TState> {
    slider: HTMLElement | null;
    componentDidMount() {
        this.setState({ valueWidth: this.slider && this.slider.offsetWidth / (this.props.max - this.props.min) });
        document.addEventListener('mouseup', this.handleMouseUp);
        window.addEventListener('resize', this.handleResize);
    }

    componentWillUnmount() {
        document.removeEventListener('mouseup', this.handleMouseUp);
        window.removeEventListener('resize', this.handleResize);
    }

    getFloatPrecision = (step: number): number => {
        const stepString = step.toString();
        const decimalIndex = stepString.indexOf('.');
        return decimalIndex >= 0 ? stepString.length - decimalIndex - 1 : 0;
    };

    roundToStep(value: number, step: number) {
        const precision = this.getFloatPrecision(step);
        let rounded = this.props.min + Math.round(Math.abs((value - this.props.min) / step)) * step;
        rounded = parseFloat(rounded.toFixed(precision));

        if (rounded < this.props.min) {
            return this.props.min;
        } else if (rounded > this.props.max) {
            return this.props.max;
        } else {
            return rounded;
        }
    }

    handleResize = () => {
        if (this.state.valueWidth * (this.props.max - this.props.min) !== this.slider?.offsetWidth) {
            this.forceUpdate();
        }
    };

    handleMouseDown = () => {
        this.setState({ isActive: true });
    };

    handleMouseUp = () => {
        this.state.isActive && this.setState({ isActive: false });
    };

    getValue = (mouseX: number, valueWidth?: number) => {
        if (mouseX < this.slider.getBoundingClientRect().left) {
            return this.props.min;
        } else if (mouseX > this.slider.getBoundingClientRect().right) {
            return this.props.max;
        } else {
            return this.roundToStep((mouseX - this.slider.getBoundingClientRect().left) / valueWidth + this.props.min, this.props.step);
        }
    };
}
