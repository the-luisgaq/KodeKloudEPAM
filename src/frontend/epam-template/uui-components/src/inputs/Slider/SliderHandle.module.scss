.container {
    pointer-events: none;

    &[data-placement^='top'] {
        .tooltip-wrapper {
            display: flex;
            flex-direction: column;
            align-items: center;
        }
    }

    &[data-placement^='bottom'] {
        .tooltip-wrapper {
            display: flex;
            flex-direction: column-reverse;
            align-items: center;
        }
    }

    &[data-placement^='left'] {
        .tooltip-wrapper {
            display: flex;
            flex-direction: row;
            align-items: center;
        }
    }

    &[data-placement^='right'] {
        .tooltip-wrapper {
            display: flex;
            flex-direction: row-reverse;
            align-items: center;
        }
    }
}
