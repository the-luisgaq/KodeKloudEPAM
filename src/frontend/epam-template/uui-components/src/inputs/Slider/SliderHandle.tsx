import * as React from 'react';
import { IHasCX, uuiElement, cx, IHasRawProps } from '@epam/uui-core';
import css from './SliderHandle.module.scss';
import { useFloating, arrow, autoUpdate } from '@floating-ui/react';
import { Portal } from '../../overlays';
import { uuiSlider } from './SliderBase';

interface SliderHandleProps extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    onUpdate(mouseX: number): void;
    onKeyDownUpdate?(type: 'right' | 'left'): void;
    handleActiveState?(isActive: boolean): void;
    tooltipContent: number;
    isActive: boolean;
    offset: number;
    showTooltip?: boolean;
}

export const SliderHandle: React.FC<SliderHandleProps> = (props) => {
    const {
        onUpdate,
        onKeyDownUpdate,
        handleActiveState,
        tooltipContent,
        isActive,
        offset: handleOffset,
        showTooltip,
        cx: propsCx,
        rawProps,
    } = props;

    const [isHovered, setIsHovered] = React.useState(false);
    const sliderHandleRef = React.useRef<HTMLDivElement | null>(null);
    const arrowRef = React.useRef<HTMLDivElement | null>(null);
    const updateTimeoutRef = React.useRef<number>(null);

    const { refs, floatingStyles } = useFloating({
        placement: 'top',
        middleware: [
            arrow({ element: arrowRef }),
        ],
        whileElementsMounted: autoUpdate,
        open: showTooltip && (isActive || isHovered),
    });

    React.useEffect(() => {
        const handleMouseMove = (e: MouseEvent) => {
            if (isActive) {
                if (updateTimeoutRef.current) {
                    window.clearTimeout(updateTimeoutRef.current);
                }
                updateTimeoutRef.current = window.setTimeout(() => {
                    onUpdate(e.clientX);
                }, 16);
            }
        };

        const handleMouseUp = () => {
            if (isActive) {
                handleActiveState?.(false);
            }
        };

        const handleMouseEnter = () => {
            setIsHovered(true);
        };

        const handleMouseLeave = () => {
            setIsHovered(false);
        };

        document.addEventListener('mousemove', handleMouseMove);
        document.addEventListener('mouseup', handleMouseUp);
        sliderHandleRef.current?.addEventListener('mouseenter', handleMouseEnter);
        sliderHandleRef.current?.addEventListener('mouseleave', handleMouseLeave);

        return () => {
            if (updateTimeoutRef.current) {
                window.clearTimeout(updateTimeoutRef.current);
            }
            document.removeEventListener('mousemove', handleMouseMove);
            document.removeEventListener('mouseup', handleMouseUp);
            sliderHandleRef.current?.removeEventListener('mouseenter', handleMouseEnter);
            sliderHandleRef.current?.removeEventListener('mouseleave', handleMouseLeave);
        };
    }, [isActive, onUpdate, handleActiveState]);

    const handleMouseDown = (e: React.MouseEvent<HTMLDivElement>): void => {
        e.preventDefault();
        handleActiveState?.(true);
    };

    const handleFocus = (e: React.FocusEvent<HTMLDivElement>): void => {
        e.preventDefault();
        handleActiveState?.(true);
        setIsHovered(true);
    };

    const handleBlur = (e: React.FocusEvent<HTMLDivElement>): void => {
        e.preventDefault();
        handleActiveState?.(false);
        setIsHovered(false);
    };

    const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
        if (e.key === 'ArrowLeft') {
            onKeyDownUpdate?.('left');
        } else if (e.key === 'ArrowRight') {
            onKeyDownUpdate?.('right');
        }
    };

    const tooltip = React.useMemo(() => {
        if (!showTooltip || (!isActive && !isHovered)) return null;

        return (
            <Portal>
                <div
                    ref={ refs.setFloating }
                    style={ floatingStyles }
                    className={ cx(propsCx, css.container, uuiElement.tooltipContainer, css.tooltipWrapper) }
                >
                    <div className={ uuiElement.tooltipBody }>{tooltipContent}</div>
                    <div ref={ arrowRef } className={ uuiElement.tooltipArrow } />
                </div>
            </Portal>
        );
    }, [showTooltip, isActive, isHovered, floatingStyles, tooltipContent, propsCx, refs.setFloating]);

    const setRefs = React.useCallback(
        (node: HTMLDivElement | null) => {
            sliderHandleRef.current = node;
            refs.setReference(node);
        },
        [refs],
    );

    return (
        <>
            <div
                role="slider"
                aria-valuenow={ tooltipContent }
                tabIndex={ 0 }
                ref={ setRefs }
                className={ cx(uuiSlider.handle, propsCx) }
                style={ { transform: `translateX(${handleOffset || 0}px)` } }
                onMouseDown={ handleMouseDown }
                onKeyDown={ handleKeyDown }
                onFocus={ handleFocus }
                onBlur={ handleBlur }
                { ...rawProps }
            />
            {tooltip}
        </>
    );
};
