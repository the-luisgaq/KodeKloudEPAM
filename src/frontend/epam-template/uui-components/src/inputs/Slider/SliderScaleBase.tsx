import * as React from 'react';
import { SliderScaleElement } from './SliderScaleElement';
import { uuiSlider } from './SliderBase';

interface SliderScaleProps<THandleOffsetValue> {
    min: number;
    max: number;
    slider: HTMLElement | null;
    valueWidth: number;
    splitAt: number;
    handleOffset: THandleOffsetValue;
    scaleUnitLabel?: string;
    renderLabel?: (value: number) => string;
}

export abstract class SliderScaleBase<THandleOffsetValue> extends React.Component<SliderScaleProps<THandleOffsetValue>> {
    abstract renderSliderScaleElements(): React.ReactElement<SliderScaleElement>[];
    getPrecision = (step: number): number => {
        const stepString = step.toString();
        const decimalIndex = stepString.indexOf('.');
        return decimalIndex >= 0 ? stepString.length - decimalIndex - 1 : 0;
    };

    generateScale = (step: number): number[] => {
        const min = this.props.min;
        const max = this.props.max;
        const precision = this.getPrecision(step);
        const count = (max - min) / step;
        const scale: number[] = [min];

        for (let i = 1; i < count; i += 1) {
            const newValue = parseFloat((min + i * step).toFixed(precision));
            scale.push(newValue);
        }

        scale.push(max);

        return scale;
    };

    render() {
        return <div className={ uuiSlider.scale }>{this.renderSliderScaleElements()}</div>;
    }
}
