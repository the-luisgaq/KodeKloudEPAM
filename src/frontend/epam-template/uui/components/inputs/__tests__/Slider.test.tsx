import React from 'react';
import { Slider } from '../Slider';
import {
    mockReactPortalsForSnapshots,
    renderSnapshotWithContextAsync,
} from '@epam/uui-test-utils';

describe('Slider', () => {
    it('should render with minimum props', async () => {
        const { mockClear } = mockReactPortalsForSnapshots();
        const tree = await renderSnapshotWithContextAsync(<Slider value={ null } onValueChange={ jest.fn } min={ 0 } max={ 50 } step={ 1 } />);
        mockClear();
        expect(tree).toMatchSnapshot();
    });

    it('should render with maximum props', async () => {
        const { mockClear } = mockReactPortalsForSnapshots();
        const tree = await renderSnapshotWithContextAsync(<Slider value={ null } onValueChange={ jest.fn } min={ 0 } max={ 50 } step={ 1 } renderLabel={ (num) => `${num}` } />);
        mockClear();
        expect(tree).toMatchSnapshot();
    });
});
