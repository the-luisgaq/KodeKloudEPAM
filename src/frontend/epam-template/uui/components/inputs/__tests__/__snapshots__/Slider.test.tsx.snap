// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render with maximum props 1`] = `
<DocumentFragment>
  <div
    class="uui-slider-container root root uui-color-neutral"
  >
    <div
      class="uui-slider"
    />
    <div
      class="uui-slider-filled"
      style="width: 0px;"
    />
    <div
      class="uui-slider-scale"
    >
      <div
        class="uui-slider-scale-dot"
        style="transform: translateX(0px);"
      />
      <div
        class="uui-slider-scale-number"
        style="transform: translateX(0px);"
      >
        0
      </div>
      <div
        class="uui-slider-scale-dot"
        style="transform: translateX(0px);"
      />
      <div
        class="uui-slider-scale-number"
        style="transform: translateX(0px);"
      >
        50
      </div>
    </div>
    <div
      aria-valuemax="50"
      aria-valuemin="0"
      class="uui-slider-handle root uui-color-neutral"
      role="slider"
      style="transform: translateX(0px);"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;

exports[`Slider should render with minimum props 1`] = `
<DocumentFragment>
  <div
    class="uui-slider-container root root uui-color-neutral"
  >
    <div
      class="uui-slider"
    />
    <div
      class="uui-slider-filled"
      style="width: 0px;"
    />
    <div
      class="uui-slider-scale"
    >
      <div
        class="uui-slider-scale-dot"
        style="transform: translateX(0px);"
      />
      <div
        class="uui-slider-scale-number"
        style="transform: translateX(0px);"
      >
        0
      </div>
      <div
        class="uui-slider-scale-dot"
        style="transform: translateX(0px);"
      />
      <div
        class="uui-slider-scale-number"
        style="transform: translateX(0px);"
      >
        50
      </div>
    </div>
    <div
      aria-valuemax="50"
      aria-valuemin="0"
      class="uui-slider-handle root uui-color-neutral"
      role="slider"
      style="transform: translateX(0px);"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
